### PR TITLE
Bug 1826186: router: add gRPC interop end-2-end tests

### PR DESCRIPTION
**Pull request by frobware**
Add e2e gRPC interop test cases for the router.

This reuses the interop test cases already defined in the grpc project[1].
The following tests are exercised:

- cancel_after_begin
- cancel_after_first_response
- client_streaming
- custom_metadata
- empty_unary
- large_unary
- ping_pong
- server_streaming
- special_status_message
- status_code_and_message
- timeout_on_sleeping_server
- unimplemented_method
- unimplemented_service

The test fixture comprises a POD with three containers:
- builder 
- server
- client-shell

The builder has copies of client.go and server.go populated from a configmap. Both these binaries are built from source during test fixture setup. They have a pinned go.{mod,sum} matching the version of gRPC currently used in origin (1.26.0). The server listens for (gRPC) client connections on h2 and h2c, and is also exposed as a service. The client-shell has access to the pre-built gRPC client binary and is used to exercise the tests on the cluster.

Three routes are created:
- passthrough (maps to server/h2)
- reencrypt (maps to server/h2)
- edge (maps to server/h2c)

The interop tests are executed against the following scenarios:
- in-cluster via the service endpoints (h2 and h2c)
- in-cluster using routes
- out-of-cluster using routes (purposefully going via the load balancer)

Note: testing via edge is currently skipped and is waiting for https://github.com/openshift/router/pull/104

[1] https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md

TODO:

- [ ] Get buy-in for building client/server.go on the cluster

### CODE REVIEW

**Comment by frobware** [comment]
A e2e run looks like: 

```console
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 cancel_after_begin
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 cancel_after_first_response
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 client_streaming
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 custom_metadata
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 empty_stream
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 empty_unary
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 large_unary
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 ping_pong
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 server_streaming
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 special_status_message
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 status_code_and_message
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 timeout_on_sleeping_server
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 unimplemented_method
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 1110 unimplemented_service
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" cancel_after_begin
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" cancel_after_first_response
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" client_streaming
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" custom_metadata
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" empty_stream
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" empty_unary
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" large_unary
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" ping_pong
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" server_streaming
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" special_status_message
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" status_code_and_message
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" timeout_on_sleeping_server
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" unimplemented_method
grpc-client -host "grpc-interop.e2e-test-router-grpc-xvj2b.svc" -port 8443 -tls -ca-cert "/etc/service-ca/service-ca.crt" unimplemented_service
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure cancel_after_begin
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure cancel_after_first_response
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure client_streaming
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure custom_metadata
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure empty_unary
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure large_unary
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure ping_pong
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure server_streaming
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure special_status_message
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure status_code_and_message
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure timeout_on_sleeping_server
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure unimplemented_method
grpc-client -host "grpc-interop-passthrough-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure unimplemented_service
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure cancel_after_begin
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure cancel_after_first_response
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure client_streaming
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure custom_metadata
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure empty_unary
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure large_unary
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure ping_pong
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure server_streaming
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure special_status_message
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure status_code_and_message
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure timeout_on_sleeping_server
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure unimplemented_method
grpc-client -host "grpc-interop-reencrypt-e2e-test-router-grpc-xvj2b.apps.amcdermo-2020-04-09-0819.gcp.devcluster.openshift.com" -port 443 -tls -insecure unimplemented_service
```

**Comment by frobware** [comment]
@openshift/openshift-team-network-edge

File: test/extended/router/grpc-interop.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+				if caCert != "" {
+					cmd = fmt.Sprintf("%s -ca-cert %q", cmd, caCert)
+				}
+				if insecure {
+					cmd = fmt.Sprintf("%s -insecure", cmd)
+				}
+				return cmd
+			}
+
+			flakyTestCases := []string{
```
[the remark this answers is not in the document]
Comment (frobware) [reply test/extended/router/grpc-interop.go:75]: This looks to be out of date; I removed `gRPCInteropTestCaseNames` in lieu of discovering the test cases available from `grpc-client -list-tests`.

**Comment by frobware** [comment]
@ironcladlou @smarterclayton i think this is good to review now. The unaddressed TODO is that this builds some Go code on in the test container. Is this a good thing? If not, we need a new home for the client/server code as an image.

**Comment by ironcladlou** [comment]
/lgtm

@smarterclayton can you please sanity check out approach here? I'll approve today if nobody objects

**Comment by ironcladlou** [comment]
/hold

Want to give just a bit more time for others to vet the general approach

**Comment by frobware** [comment]
@ironcladlou I cleaned up the `Makefile` in https://github.com/openshift/origin/pull/24856/commits/b83ccc9cc5dcb92adf0d93b9e1d0976af43604e7. It's only used in development and doesn't run as part of the general e2e origin build.

**Comment by ironcladlou** [comment]
/lgtm

If tests are passing the risk seems very low — we can always refactor.

**Comment by frobware** [comment]
/retitle Bug 1826186: router: add gRPC interop end-2-end tests

**Pull request merged**